pkg/k8s/endpointsynchronizer/cep.go

```diff
@@ -225,46 +227,67 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
+					// For json patch we don't need to perform a GET for endpoints
+
+					// If it fails it means the test from the previous patch failed
+					// so we can safely replace this node in the CNP status.
+					replaceCEPStatus := []k8s.JSONPatch{
+						{
+							OP:    "replace",
+							Path:  "/status",
+							Value: mdl,
+						},
+					}
+					var createStatusPatch []byte
+					createStatusPatch, err = json.Marshal(replaceCEPStatus)
+					if err != nil {
 						return err
 					}
-				}
-
-				// We have an object to reuse. Update and push it up. In the case of an
-				// update error, we retry in the next iteration of the controller using
-				// the copy returned by Update.
-				scopedLog.Debug("Updating CEP from local copy")
-				// Update the copy of the cep
-				mdl.DeepCopyInto(&localCEP.Status)
-				switch {
-				case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
-					localCEP, err = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(localCEP)
+					localCEP, err = ciliumClient.CiliumEndpoints(namespace).Patch(podName, types.JSONPatchType, createStatusPatch, "status")
 				default:
-					localCEP, err = ciliumClient.CiliumEndpoints(namespace).Update(localCEP)
+					// We have no localCEP copy. We need to fetch it for updates, below.
+					// This is unexpected as there should be only 1 writer per CEP, this
+					// controller, and the localCEP created on startup will be used.
+					if localCEP == nil {
+						localCEP, err = ciliumClient.CiliumEndpoints(namespace).Get(podName, meta_v1.GetOptions{})
+						switch {
+						// The CEP doesn't exist in k8s. This is unexpetected but may occur
+						// if the endpoint was removed from k8s but not yet within the agent.
+						// Mark the CEP for creation on the next controller iteration. This
+						// may never occur if the controller is stopped on Endpoint delete.
+						case err != nil && k8serrors.IsNotFound(err):
+							needInit = true
+							return err
+
+						// We cannot read the upstream CEP. needInit will cause the next
+						// iteration to delete and create the CEP. This is an unexpected
+						// situation.
+						case err != nil && k8serrors.IsInvalid(err):
+							scopedLog.WithError(err).Warn("Invalid CEP during update")
+							needInit = true
+							return nil
+
+						// A real error
+						case err != nil:
+							scopedLog.WithError(err).Error("Cannot get CEP during update")
+							return err
+						}
+					}
+
+					// We have an object to reuse. Update and push it up. In the case of an
+					// update error, we retry in the next iteration of the controller using
+					// the copy returned by Update.
+					scopedLog.Debug("Updating CEP from local copy")
+					switch {
+					case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
+						mdl.DeepCopyInto(&localCEP.Status)
+						localCEP, err = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(localCEP)
+					default:
+						mdl.DeepCopyInto(&localCEP.Status)
```

Comment:
Was there a reason we moved this line into the switch? It always still happens, right? or am I missing some way it is skipped?